File: sdk/sse/client_example.js
/**
 * SSE client example — JavaScript (browser + Node.js)
 *
 * Browser:
 *   Copy the ConnectorClient class into your frontend code.
 *   Note: browsers cannot set custom headers on EventSource.
 *   If your connector requires auth, proxy the SSE endpoint through
 *   your own backend (recommended) or disable auth for browser-only setups.
 *
 * Node.js:
 *   npm install eventsource
 *   CONNECTOR_URL=https://<host>/stream/acme_prod CONNECTOR_TOKEN=tok_xyz node client_example.js
 */

// ---------------------------------------------------------------------------
// ConnectorClient
// ---------------------------------------------------------------------------

class ConnectorClient {
  /**
   * @param {string} url   Full URL to GET /stream/{product_id}
   * @param {string} token Bearer token (leave empty if no auth required)
   */
  constructor(url, token = "") {
    this._url = url;
    this._token = token;
    this._es = null;
    this._backoff = 1000; // ms
    this._maxBackoff = 30000; // ms
    this._stopped = false;

    this.onActions = null;  // (payload) => void
    this.onSummary = null;  // (payload) => void
  }

  /** Start listening. Reconnects automatically on failure. */
  connect() {
    if (this._stopped) return;
    this._openConnection();
  }

  /** Stop listening and close the connection. */
  disconnect() {
    this._stopped = true;
    if (this._es) {
      this._es.close();
      this._es = null;
    }
  }

  _openConnection() {
    // Browser EventSource does not support custom headers.
    // In Node.js, the 'eventsource' package accepts headers via the second arg.
    const options = this._token
      ? { headers: { Authorization: `Bearer ${this._token}` } }
      : {};

    // In a browser, replace `require('eventsource')` with the global `EventSource`.
    const EventSource =
      typeof globalThis.EventSource !== "undefined"
        ? globalThis.EventSource
        : require("eventsource");

    const es = new EventSource(this._url, options);
    this._es = es;

    es.onopen = () => {
      console.log("[connector] connected");
      this._backoff = 1000; // reset on successful connection
    };

    es.onmessage = (event) => {
      this._handleMessage(event.data);
    };

    // Named 'heartbeat' events — ignore
    es.addEventListener("heartbeat", () => {});

    es.onerror = (err) => {
      es.close();
      this._es = null;
      if (this._stopped) return;

      console.warn(`[connector] connection error — retrying in ${this._backoff}ms`, err);
      setTimeout(() => this._openConnection(), this._backoff);
      this._backoff = Math.min(this._backoff * 2, this._maxBackoff);
    };
  }

  _handleMessage(data) {
    let payload;
    try {
      payload = JSON.parse(data);
    } catch (e) {
      console.warn("[connector] could not parse event data:", data.slice(0, 200));
      return;
    }

    if (payload.type === "actions" && typeof this.onActions === "function") {
      this.onActions(payload);
    } else if (payload.type === "summary" && typeof this.onSummary === "function") {
      this.onSummary(payload);
    }
  }
}

// ---------------------------------------------------------------------------
// Usage example (Node.js)
// ---------------------------------------------------------------------------

if (typeof require !== "undefined" && require.main === module) {
  const url =
    process.env.CONNECTOR_URL || "http://localhost:8080/stream/acme_prod";
  const token = process.env.CONNECTOR_TOKEN || "";

  const client = new ConnectorClient(url, token);

  client.onActions = (payload) => {
    console.log(
      `[actions] product=${payload.product_id} session=${payload.session_id} count=${payload.count}`
    );
    for (const action of payload.actions) {
      console.log(`  [${action.canonical_url}] ${action.title}`);
    }
  };

  client.onSummary = (payload) => {
    console.log(
      `[summary] product=${payload.product_id} session=${payload.session_id} replaces=${payload.replaces}`
    );
    console.log(`  ${payload.summary}`);
  };

  client.connect();
  console.log(`[connector] listening on ${url}`);
}

// ---------------------------------------------------------------------------
// Browser usage (copy the ConnectorClient class above, then):
// ---------------------------------------------------------------------------
//
// const client = new ConnectorClient("https://your-connector/stream/acme_prod");
// // Note: auth header not supported in browser EventSource — proxy if needed.
//
// client.onActions = (payload) => {
//   console.log("User actions:", payload.actions);
// };
//
// client.onSummary = (payload) => {
//   console.log("Session summary:", payload.summary);
// };
//
// client.connect();

File: sdk/sse/client_example.py
"""SSE client example — Python.

Connects to the connector's SSE endpoint and prints events as they arrive.
Reconnects automatically with exponential backoff on any connection failure.

Usage:
    pip install httpx httpx-sse
    CONNECTOR_URL=https://<host>/stream/acme_prod CONNECTOR_TOKEN=tok_xyz python client_example.py
"""

from __future__ import annotations

import json
import logging
import os
import time

import httpx
from httpx_sse import connect_sse

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
logger = logging.getLogger(__name__)

CONNECTOR_URL: str = os.getenv(
    "CONNECTOR_URL", "http://localhost:8080/stream/acme_prod"
)
CONNECTOR_TOKEN: str = os.getenv("CONNECTOR_TOKEN", "")

_BACKOFF_INITIAL = 1.0
_BACKOFF_MAX = 30.0


# ---------------------------------------------------------------------------
# Event handlers — replace these with your own logic
# ---------------------------------------------------------------------------


def on_actions(payload: dict) -> None:
    """Called for every 'actions' event."""
    logger.info(
        "product=%s session=%s — %d action(s):",
        payload.get("product_id"),
        payload.get("session_id"),
        payload.get("count", 0),
    )
    for action in payload.get("actions", []):
        logger.info("  [%s] %s", action["canonical_url"], action["title"])


def on_summary(payload: dict) -> None:
    """Called for every 'summary' event."""
    logger.info(
        "product=%s session=%s — summary (replaces %d): %s",
        payload.get("product_id"),
        payload.get("session_id"),
        payload.get("replaces", 0),
        payload.get("summary", ""),
    )


# ---------------------------------------------------------------------------
# SSE consumer with reconnection
# ---------------------------------------------------------------------------


def _process_event(event) -> None:
    """Dispatch a single SSE event to the appropriate handler."""
    if event.event == "heartbeat":
        return  # keep-alive — ignore

    try:
        payload = json.loads(event.data)
    except json.JSONDecodeError:
        logger.warning("Could not parse event data: %s", event.data[:200])
        return

    event_type = payload.get("type")
    if event_type == "actions":
        on_actions(payload)
    elif event_type == "summary":
        on_summary(payload)
    else:
        logger.debug("Unknown event type: %s", event_type)


def listen() -> None:
    """Connect to the SSE endpoint and consume events, reconnecting on failure."""
    headers = {"Accept": "text/event-stream"}
    if CONNECTOR_TOKEN:
        headers["Authorization"] = f"Bearer {CONNECTOR_TOKEN}"

    backoff = _BACKOFF_INITIAL

    while True:
        try:
            logger.info("Connecting to %s", CONNECTOR_URL)
            with httpx.Client(timeout=None) as client:
                with connect_sse(
                    client, "GET", CONNECTOR_URL, headers=headers
                ) as event_source:
                    logger.info("Connected — waiting for events")
                    backoff = _BACKOFF_INITIAL  # reset on successful connection
                    for sse in event_source.iter_sse():
                        _process_event(sse)

        except httpx.HTTPStatusError as exc:
            if exc.response.status_code in (401, 403, 404):
                logger.error(
                    "Fatal error %d — check your URL and token",
                    exc.response.status_code,
                )
                raise
            logger.warning(
                "HTTP %d — will retry in %.0fs", exc.response.status_code, backoff
            )

        except Exception as exc:
            logger.warning("Connection lost (%s) — will retry in %.0fs", exc, backoff)

        time.sleep(backoff)
        backoff = min(backoff * 2, _BACKOFF_MAX)


if __name__ == "__main__":
    listen()
